File: sprint5/kotlin/lesson7code5.kt
// Функция возвращает новый корень поддерева.
fun rotate(vertex: Vertex): Vertex {
    if (kotlin.math.abs(vertex.left.height - vertex.right.height) < 2) {
        // Вращать не надо, поддерево с вершиной vertex и так сбалансировано.
        return vertex
    } else if (vertex.left.height - vertex.right.height == -2) {
        // Нам нужны левые вращения.
        val b = vertex.right
        val R = b.right
        val C = b.left

        return if (C.height <= R.height) {
            // Нужно малое левое вращение.
            smallLeftRotation(vertex)
        } else {
            // Нужно большое левое вращение.
            bigLeftRotation(vertex)
        }
    } else if (vertex.left.height - vertex.right.height == 2) {
        // Нам нужны правые вращения.
        val b = vertex.left
        val C = b.right
        val L = b.left

        return if (C.height <= L.height) {
            // Нужно малое правое вращение.
            smallRightRotation(vertex)
        } else {
            // Нужно большое правое вращение.
            bigRightRotation(vertex)
        }
    }

    // Этот код никогда не выполняется, т.к. все ветви условий возвращают значение.
    // Но тем не менее, для компилятора этот код нужен.
    return vertex
}

File: sprint5/By lessons/lesson7code5/lesson7code5.cs
Node rotate(Node vertex) {
    if (Math.Abs(GetHeight(vertex.left) - GetHeight(vertex.right)) < 2) {
        // Вращать не надо, поддерево с вершиной vertex и так сбалансировано.    
        return vertex;
    }
    if (GetHeight(vertex.left) - GetHeight(vertex.right) == -2) {
        // Нам нужны левые вращения.
        Node b = vertex.right;
        Node R = b.right;
        Node C = b.left;

        if (GetHeight(C) <= GetHeight(R)) {
            // Нужно малое левое вращение.
            return SmallLeftRotation(vertex);
        } else {
            // Нужно большое левое вращение.
            return BigLeftRotation(vertex);
        }
    }
    if (GetHeight(vertex.left) - GetHeight(vertex.right) == 2) {
        // Нам нужны правые вращения.
        Node b = vertex.left;
        Node C = b.right;
        Node L = b.left;

        if (GetHeight(C) <= GetHeight(L)) {
            // Нужно малое правое вращение.
            return SmallRightRotation(vertex);
        } else {
            // Нужно большое правое вращение.
            return BigRightRotation(vertex);
        }
    }
}
